controllers/networkchaos/tbf/types.go

```diff
@@ -0,0 +1,252 @@
+
+func newReconciler(c client.Client, log logr.Logger, req ctrl.Request, recorder record.EventRecorder) twophase.Reconciler {
+	return twophase.Reconciler{
+		InnerReconciler: &Reconciler{
+			Client:        c,
+			EventRecorder: recorder,
+			Log:           log,
+		},
+		Client: c,
+		Log:    log,
+	}
+}
+
+// NewTwoPhaseReconciler would create Reconciler for twophase package
+func NewTwoPhaseReconciler(c client.Client, log logr.Logger, req ctrl.Request, recorder record.EventRecorder) *twophase.Reconciler {
+	r := newReconciler(c, log, req, recorder)
+	return twophase.NewReconciler(r, r.Client, r.Log)
+}
+
+// NewCommonReconciler would create Reconciler for common package
+func NewCommonReconciler(c client.Client, log logr.Logger, req ctrl.Request, recorder record.EventRecorder) *common.Reconciler {
+	r := newReconciler(c, log, req, recorder)
+	return common.NewReconciler(r, r.Client, r.Log)
+}
+
+// Apply implements the reconciler.InnerReconciler.Apply
+func (r *Reconciler) Apply(ctx context.Context, req ctrl.Request, chaos reconciler.InnerObject) error {
+	networkchaos, ok := chaos.(*v1alpha1.NetworkChaos)
+	if !ok {
+		err := errors.New("chaos is not NetworkChaos")
+		r.Log.Error(err, "chaos is not NetworkChaos", "chaos", chaos)
+		return err
+	}
+
+	pods, err := utils.SelectAndFilterPods(ctx, r.Client, &networkchaos.Spec)
+
+	if err != nil {
+		r.Log.Error(err, "failed to select and filter pods")
+		return err
+	}
+
+	err = r.applyAllPods(ctx, pods, networkchaos)
+	if err != nil {
+		return err
+	}
+
+	networkchaos.Status.Experiment.Pods = []v1alpha1.PodStatus{}
+
+	for _, pod := range pods {
+		ps := v1alpha1.PodStatus{
+			Namespace: pod.Namespace,
+			Name:      pod.Name,
+			HostIP:    pod.Status.HostIP,
+			PodIP:     pod.Status.PodIP,
+			Action:    string(networkchaos.Spec.Action),
+		}
+
+		if networkchaos.Spec.Duration != nil {
+			ps.Message = fmt.Sprintf(networkTbfActionMsg, *networkchaos.Spec.Duration)
+		}
+
+		networkchaos.Status.Experiment.Pods = append(networkchaos.Status.Experiment.Pods, ps)
+	}
+	r.Event(networkchaos, v1.EventTypeNormal, utils.EventChaosInjected, "")
+	return nil
+}
+
+func (r *Reconciler) applyAllPods(ctx context.Context, pods []v1.Pod, networkchaos *v1alpha1.NetworkChaos) error {
+	g := errgroup.Group{}
+	for index := range pods {
+		pod := &pods[index]
+
+		key, err := cache.MetaNamespaceKeyFunc(pod)
+		if err != nil {
+			return err
+		}
+		networkchaos.Finalizers = utils.InsertFinalizer(networkchaos.Finalizers, key)
+
+		g.Go(func() error {
+			return r.applyPod(ctx, pod, networkchaos)
+		})
+	}
+
+	return g.Wait()
+}
+
+func (r *Reconciler) applyPod(ctx context.Context, pod *v1.Pod, networkchaos *v1alpha1.NetworkChaos) error {
+	r.Log.Info("Try to apply tbf on pod", "namespace", pod.Namespace, "name", pod.Name)
+
+	action := string(networkchaos.Spec.Action)
+	action = strings.Title(action)
+
+	spec, ok := reflect.Indirect(reflect.ValueOf(networkchaos.Spec)).FieldByName(action).Interface().(TbfSpec)
```

Comment:
We don't need to use `reflect` here. Why don't access `networkchaos.Spec.BandwidthSpec` directly?